crates/ark/src/lsp/symbols.rs

```diff
@@ -120,13 +120,56 @@ fn is_indexable(node: &Node) -> bool {
     true
 }
 
+// Function to parse a comment and return the section level and title
+fn parse_comment_as_section(comment: &str) -> Option<(usize, String)> {
+    // Match lines starting with one or more '#' followed by some non-empty content and must end with 4 or more '-' or '#'
+    // Ensure that there's actual content between the start and the trailing symbols.
+    let comment_re = regex::Regex::new(r"^(#+)\s*([^\s#-].+?)\s*(#{4,}|-{4,}|={4,})$").unwrap();
+
+    if let Some(caps) = comment_re.captures(comment) {
+        let hashes = caps.get(1)?.as_str().len(); // Count the number of '#'
+        let title = caps.get(2)?.as_str().trim().to_string(); // Extract the title text without trailing punctuations
+        return Some((hashes, title)); // Return the level based on the number of '#' and the title
+    }
+
+    None
+}
+
 fn index_node(
     node: &Node,
     contents: &Rope,
```

Comment:
Meta comment

Looks like our `indexer`, which watches the whole "workspace" (i.e. multiple folders of files), also does something similar. i.e. see `index_comment()` there.

That indexer powers `symbols()` in this file, which powers _workspace symbols_. i.e. when you hit `CMD + T` then it pops up all the symbols in the workspace - even in closed files.

<img width="1182" alt="Screenshot 2024-10-11 at 3 17 20 PM" src="https://github.com/user-attachments/assets/4306c83a-d3d7-448e-9800-c4281e8e196f">

What we are changing here powers _document symbols_, i.e. symbols within a single document. This powers the Outline editor view we have been talking about, and it powers this Outline quick pick

<img width="1164" alt="Screenshot 2024-10-11 at 3 19 04 PM" src="https://github.com/user-attachments/assets/ea03018b-0382-4d01-bfce-f62a3f1dc80c">

---

It will eventually probably be good to merge all the infrastructure here. Ideally we'd have one set of code that we can run on a file that returns all its symbols, and then we map them into either a workspace symbol or document symbol (its essentially `indexer::index_file()` i think, but it doesn't know how to create _nested_ outlines, which is important).

I don't think this should block anything, but we should keep this in mind and open an issue to merge all of this code

Reply:
That sounds very desirable! I would really appreciate the nested outline and foldable comment functionalities. Sadly, as a student primarily working on R-based data analysis projects, I'm new to Rust and unfamiliar with the ark and LSP architecture, I'm struggling to carry this out.

Reply:
We can take care of it!